### PR TITLE
Initial draft of MVVM testing walkthrough

### PR DESCRIPTION
**Pull request by weberbry**
This post provides a basic walkthrough of my talk on testing approaches for MVVM apps.

### CODE REVIEW

File: ios/MVVM_Testing.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
+## MVVM Testing Walkthrough
+
+To better understand testing in an MVVM architected app we will walkthrough a simple app and highlight testing approaches for the related view models.
+
+This post will rely on a basic music app that includes a single screen for searching for albums from a web API. Each album in the search result will be displayed as a table view row in a SearchViewController. The table view displays each album as an AlbumSearchResultTableViewCell. The AlbumSearchResultTableViewCell displays an album title in a single label and the artist name, album release year on a second label and the album cover art.
+
+In terms of view models, the SearchViewController is backed by a SearchViewModel and each AlbumSearchResultTableViewCell is backed by a AlbumSearchResultCellViewModel.
```
Comment (alexpersian) [inline ios/MVVM_Testing.md]: "...backed by **an** Album..."

File: ios/MVVM_Testing.md [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+    init(album: Album) {
+        self.album = album
+    }
+}
+```
+
+### AlbumSearchResultCellViewModel
+
+The AlbumSearchResultCellViewModel is primarily a thin wrapper around the album model, providing some presentation logic. This type of presentation logic place to start when writing view model tests. In this instance, all public variables can be easily tested by initializing the AlbumSearchResultCellViewModel with an Album. The first code below tests the basic presentation logic in the view model.
```
Comment (alexpersian) [inline ios/MVVM_Testing.md]: "This type of presentation..." sentence reads weirdly.

File: ios/MVVM_Testing.md [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+
+	self?.albumViewModels = albums.map( {AlbumSearchResultCellViewModel(album: $0)} )
+
+    	completionHandler()
+    }
+}
+```
+The view model often acts as the first layer of abstraction between the view and the business logic, networking and data layer of an app. That said, view model tests should not be responsible for testing beyond the functionality of the view model. For instance, writing a test for the above `searchFor` method should not test the actual networking later. Instead, dependency injection should be used to inject a mock networking layer configured to simulate a functioning networking layer. As shown above, the initializer allows for injection of a mock layer. The below tests demonstrates use of the initializer to inject a mock networking layer to test the `searchFor` function. The mock networking layer simply returns a searchResults array, as would be returned by the real Networking Layer.
+
+```func testBasicAlbumSearch() {
```
Comment (alexpersian) [inline ios/MVVM_Testing.md]: Return after ```

File: ios/MVVM_Testing.md [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+```
+func albumViewModelAt(index: Int) -> AlbumSearchResultCellViewModel {
+	return albumViewModels[index]
+}
+```
+
+Testing the `albumViewModelAt` method is fairly straight forward and can be accomplished by simply creating a SearchViewModel with a known order of album elements and testing to ensure that the correct album is returned for each index, as shown below.
+
+```func testAlbumViewModelAtIndex() {
```
Comment (alexpersian) [inline ios/MVVM_Testing.md]: Return after ```

**Review by alexpersian — CHANGES_REQUESTED** [review]
Also I'd recommend surrounding the class names throughout with backticks. ie: `ClassName`

File: ios/MVVM_Testing.md [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+### Conclusion
+
+As shown in the above examples, testing view models can be a good first step to provide test coverage for presentation and other basic aspects of an apps business logic. When writing tests it is important to keep the following points in mind:
+
+- Use a view model’s public API as a guide for what tests should/could be written
+
+- Only test functionality of the view model, do not attempt to test the functionality of any dependencies of the view model, use dependency injection if needed
+
+The entire app codebase for this test app can be found [here](https://github.com/weberbry/ViewModelTestingDemo).
```
Comment (Benuuu) [inline ios/MVVM_Testing.md]: I've created a repo for MVVM examples:

https://github.com/IntrepidPursuits/MVVMExamples-ios

Lets consolidate these?

File: ios/MVVM_Testing.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    func albumViewModelAt(index: Int) -> AlbumSearchResultCellViewModel {
+        return albumViewModels[index]
+    }
+
+    func searchFor(searchTerm: String, completionHandler:@escaping () -> ()) {
+        networkingLayer.searchFor(searchTerm: searchTerm) { [weak self] (albums) in
+
+        self?.albumViewModels = albums.map( {AlbumSearchResultCellViewModel(album: $0)} )
+
+
```
Comment (Benuuu) [inline ios/MVVM_Testing.md]: Clean up some of this whitespace.  Also the indentation looks wrong for `self?.albumViewModels =...`

File: ios/MVVM_Testing.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+            completionHandler()
+        }
+    }
+
+    func setHasPreviouslyLoaded(hasPreviouslyLoaded: Bool) {
```
Comment (Benuuu) [inline ios/MVVM_Testing.md]: Where is this function used?  Can we remove it from the example here?

**Review by Benuuu — APPROVED** [review]

File: ios/MVVM_Testing.md (hunk 5 above)
Comment (weberbry) [reply ios/MVVM_Testing.md]: Added PR for the MVVMExamples-ios codebase
https://github.com/IntrepidPursuits/MVVMExamples-ios/pull/2

File: ios/MVVM_Testing.md [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+func searchFor(searchTerm: String, completionHandler:@escaping () -> ()) {
+	networkingLayer.searchFor(searchTerm: searchTerm) { [weak self] (albums) in
+
+	self?.albumViewModels = albums.map( {AlbumSearchResultCellViewModel(album: $0)} )
+
+    	completionHandler()
+    }
+}
+```
+The view model often acts as the first layer of abstraction between the view and the business logic, networking and data layer of an app. That said, view model tests should not be responsible for testing beyond the functionality of the view model. For instance, writing a test for the above `searchFor` method should not test the actual networking later. Instead, dependency injection should be used to inject a mock networking layer configured to simulate a functioning networking layer. As shown above, the initializer allows for injection of a mock layer. The below tests demonstrates use of the initializer to inject a mock networking layer to test the `searchFor` function. The mock networking layer simply returns a searchResults array, as would be returned by the real Networking Layer.
```
Comment (tzm41) [inline ios/MVVM_Testing.md]: `method should not test the actual networking later`. Did you mean `layer`?

**Review by alexpersian — APPROVED** [review]

**Pull request merged**